### DISCUSSION TITLE
how to get paths for which openat() failed with bpftrace ?

### DISCUSSION THREAD

**Discussion by vladak** [bpftrace/bpftrace#2906] [Q&A] [answered]
Using this `bpftrace` one-liner, one can get the number of failed `openat` syscalls per process:
```
bpftrace -e 'tracepoint:syscalls:sys_exit_openat /args->ret < 0/ { @[comm] = count(); }'
```
How to extend this so that the associative array contains also the path of the file for which the syscall failed ?

Using dtrace it is possible to capture the arguments in the "entry" probe into a thread local variable (using the "self" structure) and access it in the "return" probe. This does not seem to be possible with bpftrace.

**Comment by viktormalik** [comment]
You can do something similar with bpftrace, but you'd use a map instead of a thread-local variable:
```
tracepoint:syscalls:sys_enter_openat { @path[tid] = str(args.filename) }
tracepoint:syscalls:sys_exit_openat /args->ret < 0/ { @[comm, @path[tid]] = count() }
END { clear(@path) }
```
The last line just makes sure that the `@path` map is deleted upon script end and it doesn't pollute the output.

**Reply by vladak** [reply]
Thanks. In the mean time I arrived at the same solution, sans the `END` "probe" which for some reason does not work on my Ubuntu 22.04.3.

**Reply by vladak** [reply]
Is there a concept of thread-local storage in bpftrace ? My worry is that using the global associative array will have impact w.r.t. memory consumption when tracing across the system.

**Reply by viktormalik** [reply, marked as answer]
There isn't. But you can always delete the entries once you don't need them, so you would do:
```
tracepoint:syscalls:sys_exit_openat
{
  if (args->ret < 0) { @[comm, @path[tid]] = count() }
  delete(@path[tid])
}
```
for the second probe. That way, the memory footprint should be kept to minimum.
